src/control/server/mgmt_svc.go

```diff
@@ -636,69 +645,212 @@ func (svc *mgmtSvc) StorageSetFaulty(ctx context.Context, req *mgmtpb.DevStateRe
+	stopChan := make(chan struct{}, 1)
+	// select until instances stop or timeout occurs (at which point get results of each instance)
+	go func() {
+		for {
+			allStopped := true
+
+			for _, instance := range svc.harness.instances {
+				if instance.IsStarted() {
+					allStopped = false
+					continue
+				}
+			}
+
+			if allStopped {
+				break
+			}
+
+			time.Sleep(1 * time.Second)
+		}
+
+		stopChan <- struct{}{}
+	}()
+
+	select {
+	case <-stopChan:
+	case <-time.After(stopTimeout):
 	}
 
-	dresp, err := mi.CallDrpc(drpc.ModuleMgmt, drpc.MethodKillRank, req)
-	if err != nil {
-		return nil, err
+	var errored bool
+	// either all instances started or Timeout
+	for _, i := range svc.harness.instances {
+		if !i.hasSuperblock() {
+			continue // FIXME: do we need to report this?
+		}
+
+		result := &mgmtpb.RanksResp_RankResult{Rank: i.getSuperblock().Rank.Uint32()}
+		if i.IsStarted() {
+			errored = true
+			result.Status = notStoppedStatus
+		}
+		resp.Results = append(resp.Results, result)
 	}
 
-	resp := &mgmtpb.DaosResp{}
-	if err = proto.Unmarshal(dresp.Body, resp); err != nil {
-		return nil, errors.Wrap(err, "unmarshal DAOS response")
+	svc.log.Debugf("MgmtSvc.KillRanks dispatch, resp:%+v\n", *resp)
+
+	if errored {
+		return resp, errors.New("not all ranks have been killed")
+	}
+
+	return resp, nil
+}
+
+// PingRanks implements the method defined for the Management Service.
+//
+// Query data-plane all instances (DAOS system members) managed by harness to verify
+// responsiveness.
+//
+// For each instance, call over dRPC and either return error for CallDrpc err or
+// populate a RanksResp_RankResult in response. Result is either populated from
+// return from dRPC which indicates activity and Status == 0, or in the case of a timeout
+// the results status will be pingTimeoutStatus.
+func (svc *mgmtSvc) PingRanks(ctx context.Context, req *mgmtpb.RanksReq) (*mgmtpb.RanksResp, error) {
+	svc.log.Debugf("MgmtSvc.PingRanks dispatch, req:%+v\n", *req)
+
+	resp := &mgmtpb.RanksResp{}
+
+	for _, i := range svc.harness.Instances() {
+		rank, ok := validateInstanceRank(svc.log, i, req.Ranks)
+		if !ok {
+			continue
+		}
+
+		resChan := make(chan *mgmtpb.RanksResp_RankResult, 1)
+		errChan := make(chan error, 1)
+		go func() {
+			dresp, err := i.CallDrpc(drpc.ModuleMgmt, drpc.MethodPingRank, nil)
+			if err != nil {
+				errChan <- err
+			}
+			if dresp.Status != 0 {
+				errChan <- errors.Errorf("unexpected return status from PingRank dRPC (%d)",
+					int(dresp.Status))
+			}
+
+			resChan <- &mgmtpb.RanksResp_RankResult{Rank: *rank, Status: 0}
+		}()
+
+		select {
+		case err := <-errChan:
+			return nil, errors.Wrapf(err, "rank %d drpc failed", *rank)
+		case result := <-resChan:
+			resp.Results = append(resp.Results, result)
+		case <-time.After(pingTimeout):
+			resp.Results = append(resp.Results,
+				&mgmtpb.RanksResp_RankResult{Rank: *rank, Status: pingTimeoutStatus})
+		}
```

Comment:
Maybe not for this PR, but might make sense to move the timeout stuff into CallDrpc() by modifying it to accept a timeout value. Or, better, create a CallDrpcWithTimeout() which accepts a timeout in addition to the other parameters and provides this wrapper around CallDrpc().